core/trie_verify.go

```diff
@@ -0,0 +1,345 @@
+				vt.sendVerifyRequest(-1)
+			}
+		case <-vt.terminalCh:
+			return
+		}
+	}
+}
+
+// sendVerifyRequest func select at most n peers from (candidatePeers-badPeers) randomly and send verify request.
+// when n<0, send to all the peers exclude badPeers.
+func (vt *verifyTask) sendVerifyRequest(n int) {
+	var validPeers []VerifyPeer
+	candidatePeers := vt.candidatePeers.GetVerifyPeers()
+	for _, p := range candidatePeers {
+		if _, ok := vt.BadPeers[p.ID()]; !ok {
+			validPeers = append(validPeers, p)
+		}
+	}
+	// if has not valid peer, log warning.
+	if len(validPeers) == 0 {
+		log.Warn("there is no valid peer for block", vt.blockHeader.Number)
+	}
+	if n < 0 || n >= len(validPeers) {
+		for _, p := range validPeers {
+			p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+		}
+		return
+	}
+
+	// if n < len(validPeers), select n peers from validPeers randomly.
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(validPeers), func(i, j int) { validPeers[i], validPeers[j] = validPeers[j], validPeers[i] })
+	for i := 0; i < n; i++ {
+		p := validPeers[i]
+		p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+	}
+}
+
+func (vt *verifyTask) compareRootHashAndWrite(msg verifyMessage, verifyCh chan common.Hash) {
+	if msg.verifyResult.Root == vt.blockHeader.Root {
+		blockhash := msg.verifyResult.BlockHash
+		// write back to manager so that manager can cache the result and delete this task.
+		verifyCh <- blockhash
+	} else {
+		vt.BadPeers[msg.peerId] = struct{}{}
+	}
+}
+
+type VerifyPeer interface {
+	RequestRoot(blockNumber uint64, blockHash common.Hash, diffHash common.Hash) error
+	ID() string
+}
+
+type verifyPeers interface {
+	GetVerifyPeers() []VerifyPeer
+}
+
+type VerifyMode uint32
+
+const (
+	LocalVerify VerifyMode = iota
+	FullVerify
+	InsecureVerify
+	NoneVerify
+)
+
+func (mode VerifyMode) IsValid() bool {
+	return mode >= LocalVerify && mode <= NoneVerify
+}
+
+func (mode VerifyMode) String() string {
+	switch mode {
+	case LocalVerify:
+		return "local"
+	case FullVerify:
+		return "full"
+	case InsecureVerify:
+		return "insecure"
+	case NoneVerify:
+		return "none"
+	default:
+		return "unknown"
+	}
+}
+
+func (mode VerifyMode) MarshalText() ([]byte, error) {
+	switch mode {
+	case LocalVerify:
+		return []byte("local"), nil
+	case FullVerify:
+		return []byte("full"), nil
+	case InsecureVerify:
+		return []byte("insecure"), nil
+	case NoneVerify:
+		return []byte("none"), nil
+	default:
+		return nil, fmt.Errorf("unknown verify mode %d", mode)
+	}
+}
+
+func (mode *VerifyMode) UnmarshalText(text []byte) error {
+	switch string(text) {
+	case "local":
+		*mode = LocalVerify
+	case "full":
+		*mode = FullVerify
+	case "insecure":
+		*mode = InsecureVerify
+	case "none":
+		*mode = NoneVerify
+	default:
+		return fmt.Errorf(`unknown sync mode %q, want "full", "light" or "insecure"`, text)
+	}
+	return nil
+}
+
+func (mode *VerifyMode) NeedRemoteVerify() bool {
```

Comment:
Please change `func (mode *VerifyMode) NeedRemoteVerify() bool` to `func (mode VerifyMode) NeedRemoteVerify() bool`, so that you can change `func NewBlockValidator(config *params.ChainConfig, blockchain *BlockChain, engine consensus.Engine, mode *VerifyMode, peers verifyPeers) *BlockValidator ` to `func NewBlockValidator(config *params.ChainConfig, blockchain *BlockChain, engine consensus.Engine, mode VerifyMode, peers verifyPeers) *BlockValidator {`